database/src/service/read.rs

```diff
@@ -199,30 +200,45 @@ impl tower::Service<ReadRequest> for DatabaseReadHandle {
 ///
 /// This is the main entrance into all `Request` handler functions.
 /// The basic structure is:
-///
 /// 1. `Request` is mapped to a handler function
 /// 2. Handler function is called
 /// 3. [`Response`] is sent
 fn map_request(
-    _permit: OwnedSemaphorePermit,   // Permit for this request
-    env: Arc<ConcreteEnv>,           // Access to the database
-    request: ReadRequest,            // The request we must fulfill
+    _permit: OwnedSemaphorePermit, // Permit for this request, dropped at end of function
+    env: Arc<ConcreteEnv>,         // Access to the database
+    request: ReadRequest,          // The request we must fulfill
     response_sender: ResponseSender, // The channel we must send the response back to
 ) {
     /* TODO: pre-request handling, run some code for each request? */
-
-    match request {
-        ReadRequest::Example1 => example_handler_1(env, response_sender),
-        ReadRequest::Example2(x) => example_handler_2(env, response_sender, x),
-        ReadRequest::Example3(x) => example_handler_3(env, response_sender, x),
-    }
+    use ReadRequest as R;
+
+    let response = match request {
+        R::BlockExtendedHeader(block) => block_extended_header(&env, block),
+        R::BlockHash(block) => block_hash(&env, block),
+        R::BlockExtendedHeaderInRange(range) => block_extended_header_in_range(&env, range),
+        R::ChainHeight => chain_height(&env),
+        R::GeneratedCoins => generated_coins(&env),
+        R::Outputs(map) => outputs(&env, map),
+        R::NumberOutputsWithAmount(vec) => number_outputs_with_amount(&env, vec),
+        R::CheckKIsNotSpent(set) => check_k_is_not_spent(&env, set),
+        R::BlockBatchInRange(range) => block_batch_in_range(&env, range),
+    };
+
+    // TODO: what do we do if this errors?
+    response_sender.send(response).unwrap();
 
     /* TODO: post-request handling, run some code for each request? */
```

Comment:
Maybe some performance counter ?
- an atomic counter of database requested handled since uptime
- some timer for measuring how much time the request take ? Maybe tracing would be enough for this, but I remember they postpone some of their writing.

Reply:
We could instrument manually and make `cuprate_database::service::init()` return some shared state like:
```rust
use std::time::Instant;

struct DatabaseStats {
    read_requests: SomeList<Instant>, // .len() == total requests
    write_requests: SomeList<Instant>,
    init_instant: Instant,
    other_stats: [...],
}

pub fn init(config: Config) ->
    Result<(
        DatabaseReadHandle,
        DatabaseWriteHandle,
        Arc<DatabaseStats> // todo: make this read-only
    ), InitError>;
```
but I think `tracing` will mostly cover everything and be better? I'm not sure.

Reply:
yeah tracing should handle this